{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up Julia processes and output dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "using Distributed\n",
    "addprocs(7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@everywhere begin\n",
    "    using Catalyst, DifferentialEquations, Distributions, CairoMakie\n",
    "    using DataFrames\n",
    "    using Parquet2\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the output directory if it does not exist\n",
    "outdir = \"$(@__DIR__)/../../output/modeling/julia_missplicing_simulations\"\n",
    "mkpath(outdir)\n",
    "\n",
    "# Make output dataframe\n",
    "data = DataFrame(copynum=Int[], reg_gene=Int[], unreg_gene=Int[], design=String[], moi=Float64[], risc=Int[])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define reaction networks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Constituent reaction networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@everywhere begin\n",
    "    \n",
    "    # RISC-mediated knockdown of the regulated gene, `regulated_mRNA`\n",
    "    dicer_risc_rn = @reaction_network dicer_RISC_complex begin\n",
    "        @parameters begin\n",
    "            r_dicer=1e-3            # pre-miRNA processing rate constant [1/s]\n",
    "            δ_mi=2.88e-4            # miRNA degradation rate constant [1/s]\n",
    "            k_miRNA_bind=1e-5       # miRNA-RISC binding rate constant [1/mol*s]\n",
    "            k_miRNA_deg=2.16e-5  # miRNA-RISC unbinding rate constant [1/s]\n",
    "            k_mRNA_bind=1.84e-6     # RISC complex formation rate constant[1/mol*s]\n",
    "            k_mRNA_unbind=0.303     # RISC complex deformation rate constant [1/s]\n",
    "            k_deg=7e-3              # RISC-mediated mRNA degradation rate constant [1/s]\n",
    "        end\n",
    "        # dicer\n",
    "        r_dicer, pre_miRNA --> miRNA\n",
    "        # degradation\n",
    "        δ_mi, miRNA --> ∅\n",
    "        # Knockdown\n",
    "        k_miRNA_bind, risc + miRNA --> risc_miRNA\n",
    "        k_miRNA_deg, risc_miRNA --> risc\n",
    "        (k_mRNA_bind, k_mRNA_unbind), risc_miRNA + regulated_mRNA <--> risc_miRNA_mRNA\n",
    "        k_deg, risc_miRNA_mRNA --> risc_miRNA\n",
    "    end\n",
    "\n",
    "    # RISC-mediated knockdown of the immature unspliced transcript, 'immature_mRNA'\n",
    "    immature_risc_rn = @reaction_network immature_RISC_complex begin\n",
    "        @parameters begin\n",
    "            k_mRNA_bind=1.84e-6     # RISC complex formation rate constant[1/mol*s]\n",
    "            k_mRNA_unbind=0.303     # RISC complex deformation rate constant [1/s]\n",
    "            k_deg=7e-3              # RISC-mediated mRNA degradation rate constant [1/s]\n",
    "        end\n",
    "        # Knockdown\n",
    "        (k_mRNA_bind, k_mRNA_unbind), risc_miRNA + immature_mRNA <--> risc_miRNA_im_mRNA\n",
    "        k_deg, risc_miRNA_im_mRNA --> risc_miRNA\n",
    "    end\n",
    "\n",
    "    # Regulated gene dynamics\n",
    "    regulated_gene_rn = @reaction_network regulated_gene_dynamics begin\n",
    "        @parameters begin\n",
    "            δ_m=2.88e-4 # Mature mRNA degradation rate constant [1/s]\n",
    "            δ_p=9.67e-5 # Protein degradation rate constant [1/s] \n",
    "            α_p=3.33e-4 # Translation rate constant [1/s]\n",
    "            ζ=0.0       # Translation efficiency for RISC complex [unitless], Range: 0 to 1\n",
    "        end\n",
    "        # Production of protein\n",
    "        α_p, regulated_mRNA --> protein + regulated_mRNA\n",
    "        α_p, control_mRNA --> control_protein + control_mRNA\n",
    "        α_p * ζ, risc_miRNA_mRNA --> protein + risc_miRNA_mRNA\n",
    "        # Degradation of products\n",
    "        δ_m, regulated_mRNA --> ∅\n",
    "        δ_m, control_mRNA --> ∅\n",
    "        δ_p, protein --> ∅\n",
    "        δ_p, control_protein --> ∅\n",
    "    end\n",
    "\n",
    "    # Immature mRNA translation dynamics\n",
    "    immature_translation_correct_rn = @reaction_network immature_correct_translation begin\n",
    "        @parameters begin\n",
    "            δ_m=2.88e-4 # Mature mRNA degradation rate constant [1/s]\n",
    "            δ_p=9.67e-5 # Protein degradation rate constant [1/s] \n",
    "            α_p=3.33e-4 # Translation rate constant [1/s]\n",
    "            ζ=0.0       # Translation efficiency for RISC complex [unitless], Range: 0 to 1\n",
    "        end\n",
    "        # Production of protein\n",
    "        α_p, immature_mRNA --> protein + immature_mRNA\n",
    "        α_p * ζ, risc_miRNA_im_mRNA --> protein + risc_miRNA_im_mRNA\n",
    "    end\n",
    "    immature_translation_incorrect_rn = @reaction_network immature_incorrect_translation begin\n",
    "        @parameters begin\n",
    "            δ_m=2.88e-4 # Mature mRNA degradation rate constant [1/s]\n",
    "            δ_p=9.67e-5 # Protein degradation rate constant [1/s] \n",
    "            α_p=3.33e-4 # Translation rate constant [1/s]\n",
    "            ζ=0.0       # Translation efficiency for RISC complex [unitless], Range: 0 to 1\n",
    "        end\n",
    "        # Production of protein\n",
    "        α_p, immature_mRNA --> immature_mRNA + incorrect_protein\n",
    "        α_p * ζ, risc_miRNA_im_mRNA --> incorrect_protein + risc_miRNA_im_mRNA\n",
    "        # Degradation of products\n",
    "        δ_p, incorrect_protein --> ∅\n",
    "    end\n",
    "\n",
    "    # Production of a regulated transcript with its own miRNA intron\n",
    "    intronic_regulated_gene_rn = @reaction_network intronic_regulated_expression begin\n",
    "        @parameters begin\n",
    "            α_im=4.67e-2        # Transcription rate constant [1/s]\n",
    "            δ_im=2.88e-4        # Immature mRNA degradation rate constant [1/s]\n",
    "            r_splicing=2.0e-3   # Splicing rate constant [1/s]\n",
    "            r_drosha=1e-2       # pri-miRNA processing rate constant [1/s]\n",
    "        end\n",
    "        # Transcription\n",
    "        α_im, gene --> immature_mRNA + gene\n",
    "        α_im, gene --> control_mRNA + gene\n",
    "        # Degradation of products\n",
    "        δ_im, immature_mRNA --> ∅  \n",
    "        # Splicing\n",
    "        r_splicing, immature_mRNA --> regulated_mRNA + pri_miRNA\n",
    "        # miRNA processing: drosha\n",
    "        r_drosha, pri_miRNA --> pre_miRNA\n",
    "    end\n",
    "\n",
    "    # Production of an regulated transcript with its own miRNA intron\n",
    "    missplicing_rn = @reaction_network missplicing begin\n",
    "        @parameters begin\n",
    "            r_mRNA_only=0.0   # Splicing rate constant [1/s]\n",
    "            r_miR_only=0.0   # Splicing rate constant [1/s]\n",
    "        end\n",
    "        # Splicing\n",
    "        r_mRNA_only, immature_mRNA --> regulated_mRNA\n",
    "        r_miR_only, immature_mRNA --> pri_miRNA\n",
    "    end\n",
    "\n",
    "    # Production of a regulated gene transcript, without an internal intron\n",
    "    nonintronic_regulated_gene_rn = @reaction_network gene_ts_expression begin\n",
    "        @parameters begin\n",
    "            α_m=4.67e-2     # Transcription rate constant [1/s]\n",
    "        end\n",
    "        α_m, gene --> regulated_mRNA + gene\n",
    "        α_m, gene --> control_mRNA + gene\n",
    "    end\n",
    "\n",
    "    # Production of an unregulated transcript with the miRNA intron\n",
    "    intronic_unregulated_gene_rn = @reaction_network intronic_unregulated_expression begin\n",
    "        @parameters begin\n",
    "            α_im=4.67e-2       # Transcription rate constant [1/s]\n",
    "            α_p=3.33e-4        # Translation rate constant [1/s]\n",
    "            δ_m=2.88e-4        # Mature mRNA degradation rate constant [1/s]\n",
    "            δ_p=9.67e-5        # Protein degradation rate constant [1/s] \n",
    "            δ_im=2.88e-4       # Immature mRNA degradation rate constant gene 2 [1/s]\n",
    "            r_splicing=2.0e-3   # Splicing rate constant [1/s]\n",
    "            r_drosha=1e-2       # pri-miRNA processing rate constant [1/s]\n",
    "        end\n",
    "        # Transcription\n",
    "        α_im, miRNA_gene --> immature_mRNA + miRNA_gene\n",
    "        # Degradation of products\n",
    "        δ_im, immature_mRNA --> ∅\n",
    "        # Splicing\n",
    "        r_splicing, immature_mRNA --> pri_miRNA\n",
    "        # miRNA processing: drosha\n",
    "        r_drosha, pri_miRNA --> pre_miRNA\n",
    "    end\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overall design reaction networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@everywhere begin\n",
    "    common_rn = extend(dicer_risc_rn, regulated_gene_rn)\n",
    "    CDS_intron_rn = extend(extend(intronic_regulated_gene_rn, immature_risc_rn), immature_translation_incorrect_rn)\n",
    "\n",
    "    base_iFFL = extend(extend(CDS_intron_rn, common_rn), missplicing_rn)\n",
    "    base_params = [:r_splicing => 2.0e-3, :r_mRNA_only => 0.0, :r_miR_only => 0.0]\n",
    "\n",
    "    more_mRNA_params = [:r_splicing => 2.0e-3, :r_mRNA_only => 2.0e-3, :r_miR_only => 0.0]\n",
    "\n",
    "    more_miR_params = [:r_splicing => 2.0e-3, :r_mRNA_only => 0.0, :r_miR_only => 2.0e-3]\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species(base_iFFL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Base iFFL simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@everywhere risc_vals = [10000]\n",
    "@everywhere moi_vals = [0.3, 3, 10]\n",
    "\n",
    "for k=1:lastindex(risc_vals)\n",
    "    #iterate through all initial values of risc\n",
    "    for j=1:lastindex(moi_vals)\n",
    "        #iterate through all moi values for Poisson distribution\n",
    "        \n",
    "        @everywhere begin\n",
    "            #define copy number distribution\n",
    "            CN_distribution = truncated(Poisson(moi_vals[$j]), lower=1.0)\n",
    "\n",
    "            #define method to resample from copy number distribution\n",
    "            function prob_func(prob, i, repeat)\n",
    "                remake(prob, u0 = [:gene => rand(CN_distribution), :risc => risc_vals[$k], :immature_mRNA => 0, :control_mRNA => 0, :miRNA => 0, \n",
    "                    :regulated_mRNA => 0, :protein => 0, :control_protein => 0, :pri_miRNA => 0, :pre_miRNA => 0, :risc_miRNA => 0, :risc_miRNA_mRNA => 0,\n",
    "                    :risc_miRNA_im_mRNA => 0, :incorrect_protein => 0])\n",
    "            end\n",
    "\n",
    "            #define simulation output function\n",
    "            function output_func(sol, i)\n",
    "                sol.u[end,:], false\n",
    "            end\n",
    "\n",
    "            #define initial values for simulation\n",
    "            u0 = [:gene => rand(CN_distribution), :risc => risc_vals[$k], :immature_mRNA => 0, :control_mRNA => 0, :miRNA => 0, \n",
    "                    :regulated_mRNA => 0, :protein => 0, :control_protein => 0, :pri_miRNA => 0, :pre_miRNA => 0, :risc_miRNA => 0, :risc_miRNA_mRNA => 0,\n",
    "                    :risc_miRNA_im_mRNA => 0, :incorrect_protein => 0]\n",
    "\n",
    "            #define tspan\n",
    "            tspan = (0., 100000.)\n",
    "            \n",
    "            #define jump problem for desired reaction network\n",
    "            jprob = JumpProblem(base_iFFL, DiscreteProblem(base_iFFL, u0, tspan, base_params), SortingDirect(), save_positions=(false,false))\n",
    "\n",
    "            #define ensemble problem with functions from above\n",
    "            eprob = EnsembleProblem(jprob, prob_func=prob_func, output_func=output_func, safetycopy=false)\n",
    "\n",
    "        end\n",
    "\n",
    "        #solve given ensemble problem\n",
    "        solns = solve(eprob, SSAStepper(), EnsembleSplitThreads(); trajectories=10000)\n",
    "\n",
    "        #extract desired data from output\n",
    "        reg_vec = zeros(10000)\n",
    "        unreg_vec = zeros(10000)\n",
    "        cn_vec = zeros(10000)\n",
    "        for n = 1:lastindex(solns)\n",
    "            reg_vec[n] = solns[n][1][4]\n",
    "            unreg_vec[n] = solns[n][1][6]\n",
    "            cn_vec[n] = solns[n][1][14]\n",
    "        end\n",
    "\n",
    "        df = DataFrame(copynum=cn_vec, reg_gene=reg_vec, unreg_gene=unreg_vec, design=\"Base\", moi=moi_vals[j], risc=risc_vals[k])\n",
    "        append!(data,df)\n",
    "\n",
    "        f = Figure()\n",
    "        ax1 = Axis(f[1,1],title=\"Base parameters: \"*string(risc_vals[k])*\" RISC, \"*string(moi_vals[j])*\" MOI\", xlabel = \"Copy Number\", ylabel = \"Protein Expression\", limits = (nothing, nothing, 0, 10000)\n",
    "            )\n",
    "        s1 = scatter!(ax1, df.copynum, df.unreg_gene, color=(:gray, 0.5), label=\"unregulated gene\")\n",
    "        s2 = scatter!(ax1, df.copynum, df.reg_gene, color=(:blue, 0.5), label=\"regulated gene\")\n",
    "        axislegend(position=:lt)\n",
    "        ax2 = Axis(f[1,2],limits = (nothing, nothing, 0, 10000))\n",
    "        linkyaxes!(ax1,ax2)\n",
    "        d1 = density!(df.unreg_gene, color=(:gray, 0.5), direction = :y)\n",
    "        d2 = density!(df.reg_gene, color=(:blue, 0.5), direction = :y)\n",
    "        xlims!(ax2, low = 0)\n",
    "        hidedecorations!(ax2, grid = false)\n",
    "        colsize!(f.layout, 1, Auto(3))\n",
    "        save(\"$outdir/base_\"*string(risc_vals[k])*\"risc_\"*string(moi_vals[j])*\"moi.svg\",f)\n",
    "    end\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Parquet2.writefile(\"$outdir/stochastic_sims_base.gzip\", data; compression_codec=:gzip)\n",
    "Parquet2.writefile(\"$outdir/stochastic_sims.gzip\", data; compression_codec=:gzip)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_base = copy(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# More mRNA simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_more_mRNA = DataFrame(copynum=Int[], reg_gene=Int[], unreg_gene=Int[], design=String[], moi=Float64[], risc=Int[])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@everywhere risc_vals = [10000]\n",
    "@everywhere moi_vals = [0.3, 3, 10]\n",
    "\n",
    "for k=1:lastindex(risc_vals)\n",
    "    #iterate through all initial values of risc\n",
    "    for j=1:lastindex(moi_vals)\n",
    "        #iterate through all moi values for Poisson distribution\n",
    "        \n",
    "        @everywhere begin\n",
    "            #define copy number distribution\n",
    "            CN_distribution = truncated(Poisson(moi_vals[$j]), lower=1.0)\n",
    "\n",
    "            #define method to resample from copy number distribution\n",
    "            function prob_func(prob, i, repeat)\n",
    "                remake(prob, u0 = [:gene => rand(CN_distribution), :risc => risc_vals[$k], :immature_mRNA => 0, :control_mRNA => 0, :miRNA => 0, \n",
    "                    :regulated_mRNA => 0, :protein => 0, :control_protein => 0, :pri_miRNA => 0, :pre_miRNA => 0, :risc_miRNA => 0, :risc_miRNA_mRNA => 0,\n",
    "                    :risc_miRNA_im_mRNA => 0, :incorrect_protein => 0])\n",
    "            end\n",
    "\n",
    "            #define simulation output function\n",
    "            function output_func(sol, i)\n",
    "                sol.u[end,:], false\n",
    "            end\n",
    "\n",
    "            #define initial values for simulation\n",
    "            u0 = [:gene => rand(CN_distribution), :risc => risc_vals[$k], :immature_mRNA => 0, :control_mRNA => 0, :miRNA => 0, \n",
    "                    :regulated_mRNA => 0, :protein => 0, :control_protein => 0, :pri_miRNA => 0, :pre_miRNA => 0, :risc_miRNA => 0, :risc_miRNA_mRNA => 0,\n",
    "                    :risc_miRNA_im_mRNA => 0, :incorrect_protein => 0]\n",
    "\n",
    "            #define tspan\n",
    "            tspan = (0., 100000.)\n",
    "            \n",
    "            #define jump problem for desired reaction network\n",
    "            jprob = JumpProblem(base_iFFL, DiscreteProblem(base_iFFL, u0, tspan, more_mRNA_params), SortingDirect(), save_positions=(false,false))\n",
    "\n",
    "            #define ensemble problem with functions from above\n",
    "            eprob = EnsembleProblem(jprob, prob_func=prob_func, output_func=output_func, safetycopy=false)\n",
    "\n",
    "        end\n",
    "\n",
    "        #solve given ensemble problem\n",
    "        solns = solve(eprob, SSAStepper(), EnsembleSplitThreads(); trajectories=10000)\n",
    "\n",
    "        #extract desired data from output\n",
    "        reg_vec = zeros(10000)\n",
    "        unreg_vec = zeros(10000)\n",
    "        cn_vec = zeros(10000)\n",
    "        for n = 1:lastindex(solns)\n",
    "            reg_vec[n] = solns[n][1][4]\n",
    "            unreg_vec[n] = solns[n][1][6]\n",
    "            cn_vec[n] = solns[n][1][14]\n",
    "        end\n",
    "\n",
    "        df = DataFrame(copynum=cn_vec, reg_gene=reg_vec, unreg_gene=unreg_vec, design=\"more mRNA\", moi=moi_vals[j], risc=risc_vals[k])\n",
    "        append!(data,df)\n",
    "        append!(data_more_mRNA,df)\n",
    "\n",
    "        f = Figure()\n",
    "        ax1 = Axis(f[1,1],title=\"More mRNA: \"*string(risc_vals[k])*\" RISC, \"*string(moi_vals[j])*\" MOI\", xlabel = \"Copy Number\", ylabel = \"Protein Expression\", limits = (nothing, nothing, 0, 10000)\n",
    "            )\n",
    "        s1 = scatter!(ax1, df.copynum, df.unreg_gene, color=(:gray, 0.5), label=\"unregulated gene\")\n",
    "        s2 = scatter!(ax1, df.copynum, df.reg_gene, color=(:blue, 0.5), label=\"regulated gene\")\n",
    "        axislegend(position=:lt)\n",
    "        ax2 = Axis(f[1,2],limits = (nothing, nothing, 0, 10000))\n",
    "        linkyaxes!(ax1,ax2)\n",
    "        d1 = density!(df.unreg_gene, color=(:gray, 0.5), direction = :y)\n",
    "        d2 = density!(df.reg_gene, color=(:blue, 0.5), direction = :y)\n",
    "        xlims!(ax2, low = 0)\n",
    "        hidedecorations!(ax2, grid = false)\n",
    "        colsize!(f.layout, 1, Auto(3))\n",
    "        save(\"$outdir/more_mRNA_\"*string(risc_vals[k])*\"risc_\"*string(moi_vals[j])*\"moi.svg\",f)\n",
    "    end\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Parquet2.writefile(\"$outdir/stochastic_sims.gzip\", data; compression_codec=:gzip)\n",
    "Parquet2.writefile(\"$outdir/stochastic_sims_more_mRNA.gzip\", data_more_mRNA; compression_codec=:gzip)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# More miRNA simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_more_miR = DataFrame(copynum=Int[], reg_gene=Int[], unreg_gene=Int[], design=String[], moi=Float64[], risc=Int[])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@everywhere risc_vals = [10000]\n",
    "@everywhere moi_vals = [0.3, 3, 10]\n",
    "\n",
    "for k=1:lastindex(risc_vals)\n",
    "    #iterate through all initial values of risc\n",
    "    for j=1:lastindex(moi_vals)\n",
    "        #iterate through all moi values for Poisson distribution\n",
    "        \n",
    "        @everywhere begin\n",
    "            #define copy number distribution\n",
    "            CN_distribution = truncated(Poisson(moi_vals[$j]), lower=1.0)\n",
    "\n",
    "            #define method to resample from copy number distribution\n",
    "            function prob_func(prob, i, repeat)\n",
    "                remake(prob, u0 = [:gene => rand(CN_distribution), :risc => risc_vals[$k], :immature_mRNA => 0, :control_mRNA => 0, :miRNA => 0, \n",
    "                    :regulated_mRNA => 0, :protein => 0, :control_protein => 0, :pri_miRNA => 0, :pre_miRNA => 0, :risc_miRNA => 0, :risc_miRNA_mRNA => 0,\n",
    "                    :risc_miRNA_im_mRNA => 0, :incorrect_protein => 0])\n",
    "            end\n",
    "\n",
    "            #define simulation output function\n",
    "            function output_func(sol, i)\n",
    "                sol.u[end,:], false\n",
    "            end\n",
    "\n",
    "            #define initial values for simulation\n",
    "            u0 = [:gene => rand(CN_distribution), :risc => risc_vals[$k], :immature_mRNA => 0, :control_mRNA => 0, :miRNA => 0, \n",
    "                    :regulated_mRNA => 0, :protein => 0, :control_protein => 0, :pri_miRNA => 0, :pre_miRNA => 0, :risc_miRNA => 0, :risc_miRNA_mRNA => 0,\n",
    "                    :risc_miRNA_im_mRNA => 0, :incorrect_protein => 0]\n",
    "\n",
    "            #define tspan\n",
    "            tspan = (0., 100000.)\n",
    "            \n",
    "            #define jump problem for desired reaction network\n",
    "            jprob = JumpProblem(base_iFFL, DiscreteProblem(base_iFFL, u0, tspan, more_miR_params), SortingDirect(), save_positions=(false,false))\n",
    "\n",
    "            #define ensemble problem with functions from above\n",
    "            eprob = EnsembleProblem(jprob, prob_func=prob_func, output_func=output_func, safetycopy=false)\n",
    "\n",
    "        end\n",
    "\n",
    "        #solve given ensemble problem\n",
    "        solns = solve(eprob, SSAStepper(), EnsembleSplitThreads(); trajectories=10000)\n",
    "\n",
    "        #extract desired data from output\n",
    "        reg_vec = zeros(10000)\n",
    "        unreg_vec = zeros(10000)\n",
    "        cn_vec = zeros(10000)\n",
    "        for n = 1:lastindex(solns)\n",
    "            reg_vec[n] = solns[n][1][4]\n",
    "            unreg_vec[n] = solns[n][1][6]\n",
    "            cn_vec[n] = solns[n][1][14]\n",
    "        end\n",
    "\n",
    "        df = DataFrame(copynum=cn_vec, reg_gene=reg_vec, unreg_gene=unreg_vec, design=\"more miR\", moi=moi_vals[j], risc=risc_vals[k])\n",
    "        append!(data,df)\n",
    "        append!(data_more_miR,df)\n",
    "\n",
    "        f = Figure()\n",
    "        ax1 = Axis(f[1,1],title=\"More miRNA: \"*string(risc_vals[k])*\" RISC, \"*string(moi_vals[j])*\" MOI\", xlabel = \"Copy Number\", ylabel = \"Protein Expression\", limits = (nothing, nothing, 0, 10000)\n",
    "            )\n",
    "        s1 = scatter!(ax1, df.copynum, df.unreg_gene, color=(:gray, 0.5), label=\"unregulated gene\")\n",
    "        s2 = scatter!(ax1, df.copynum, df.reg_gene, color=(:blue, 0.5), label=\"regulated gene\")\n",
    "        axislegend(position=:lt)\n",
    "        ax2 = Axis(f[1,2],limits = (nothing, nothing, 0, 10000))\n",
    "        linkyaxes!(ax1,ax2)\n",
    "        d1 = density!(df.unreg_gene, color=(:gray, 0.5), direction = :y)\n",
    "        d2 = density!(df.reg_gene, color=(:blue, 0.5), direction = :y)\n",
    "        xlims!(ax2, low = 0)\n",
    "        hidedecorations!(ax2, grid = false)\n",
    "        colsize!(f.layout, 1, Auto(3))\n",
    "        save(\"$outdir/more_miRNA_\"*string(risc_vals[k])*\"risc_\"*string(moi_vals[j])*\"moi.svg\",f)\n",
    "    end\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Parquet2.writefile(\"$outdir/stochastic_sims.gzip\", data; compression_codec=:gzip)\n",
    "Parquet2.writefile(\"$outdir/stochastic_sims_more_miRNA.gzip\", data_more_miR; compression_codec=:gzip)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.4",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
